analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -93,61 +91,66 @@ public override void ExecutionCompleted()
     {
         List<Location> locations = new();
-        IOperation currentStart = null;
-        IOperation currentEnd = null;
-        var unreachable = UnreachableOperations(block, conditionValue).Where(x => !IsIgnoredLocation(x.Syntax)).ToHashSet();
-        foreach (var operation in unreachable.Concat(reached).OrderBy(x => x.Syntax.SpanStart))
+        var unreachable = UnreachableOperations(block, conditionValue);
+        var currentStart = conditionSyntax.Span.End;
+        var reachedNodes = reached.Select(x => x.Syntax).Where(x => x.SpanStart > conditionSyntax.Span.End).OrderBy(x => x.SpanStart);
+
+        foreach (var reachedNode in reachedNodes)
         {
-            if (unreachable.Contains(operation))
-            {
-                currentStart ??= operation;
-                currentEnd = operation;
-            }
-            else
+            if (AddLocation(reachedNode.SpanStart))
             {
-                AddCurrent();
+                currentStart = reachedNode.Span.End;
             }
         }
-        AddCurrent();
+        AddLocation(int.MaxValue);
         return locations;
 
-        void AddCurrent()
+        bool AddLocation(int end)
         {
-            if (currentStart is not null)
+            var nodes = unreachable.Where(x => x.SpanStart > currentStart && x.Span.End < end);
+            if (nodes.Any())
             {
-                locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
-                currentStart = null;
+                var first = nodes.OrderBy(x => x.SpanStart).First();
+                var last = nodes.OrderBy(x => x.Span.End).Last();
+                locations.Add(first.CreateLocation(last));
+                return true;
             }
+            return false;
         }
     }
 
-    private IEnumerable<IOperation> UnreachableOperations(BasicBlock block, bool conditionValue)
+    private IEnumerable<SyntaxNode> UnreachableOperations(BasicBlock block, bool conditionValue)
     {
         if (block.SuccessorBlocks.Distinct().Count() != 2)
         {
-            return Enumerable.Empty<IOperation>();
+            return Enumerable.Empty<SyntaxNode>();
         }
         HashSet<BasicBlock> reachable = new() { block };
         HashSet<BasicBlock> unreachable = new();
 
         var conditionalIsRechable = (block.ConditionKind == ControlFlowConditionKind.WhenTrue) == conditionValue;
         Traverse(conditionalIsRechable ? block.ConditionalSuccessor : block.FallThroughSuccessor, reachable, new List<BasicBlock>());
         Traverse(conditionalIsRechable ? block.FallThroughSuccessor : block.ConditionalSuccessor, unreachable, reachable);
-        return unreachable.SelectMany(x => x.OperationsAndBranchValue).Except(reached);
+        return unreachable
+            .SelectMany(x => x.OperationsAndBranchValue)
+            .Except(reached)
+            .SelectMany(x => x.DescendantsAndSelf().Select(x => x.Syntax))
+            .ToList();
```

Comment:
I think it's better here to transform the HashSet to list first and then run the query (since list is continuous in memory) - it should be faster to do the first enumeration.

I think there's no need to transform it to a list on line 153 in any case (you just need to return IEnumerable).

Reply:
I don't think transforming it first makes any difference.
The transform at the end is there so it will not be evaluated multiple times.
I updated the method signature to make the intent more clear.